File: zig/src/starknet_client/serialization.zig
const std = @import("std");
const core = @import("../core/root.zig");
const she = @import("../she/root.zig");
const tongo_ops = @import("../tongo/operations.zig");

pub const Error = error{
    InvalidPoint,
    InvalidInput,
    OutOfMemory,
};

pub fn serializeProjectivePoint(point: core.ProjectivePoint) Error![2]core.Felt {
    const affine = point.toAffine() orelse return Error.InvalidPoint;
    return .{ affine.x, affine.y };
}

pub fn deserializeProjectivePoint(x: core.Felt, y: core.Felt) Error!core.ProjectivePoint {
    if (!core.AffinePoint.new(x, y).isOnCurve()) return Error.InvalidPoint;
    return core.ProjectivePoint.fromAffine(x, y);
}

pub fn serializePoeProof(proof: she.types.PoeProof) Error![3]core.Felt {
    const a = try serializeProjectivePoint(proof.a);
    return .{ a[0], a[1], proof.s };
}

pub fn serializePoe2Proof(proof: she.types.Poe2Proof) Error![4]core.Felt {
    const a = try serializeProjectivePoint(proof.a);
    return .{ a[0], a[1], proof.s1, proof.s2 };
}

pub fn serializeElgamalProof(proof: she.types.ElGamalProof) Error![6]core.Felt {
    const al = try serializeProjectivePoint(proof.al);
    const ar = try serializeProjectivePoint(proof.ar);
    return .{ al[0], al[1], ar[0], ar[1], proof.sb, proof.sr };
}

pub fn u128ToU256(value: u128) [2]core.Felt {
    return .{
        feltFromU128(value),
        core.Felt.ZERO,
    };
}

pub fn u256ToU128(low: core.Felt, high: core.Felt) Error!u128 {
    if (!core.Felt.eql(high, core.Felt.ZERO)) return Error.InvalidInput;
    const bytes = low.toBytesBe();
    return std.mem.readInt(u128, bytes[16..32], .big);
}

pub fn serializeAeBalance(
    ciphertext_bytes: [64]u8,
    nonce_bytes: [24]u8,
) [6]core.Felt {
    const ct_felts = bytesToU512(ciphertext_bytes);

    var nonce_padded: [32]u8 = [_]u8{0} ** 32;
    @memcpy(nonce_padded[0..24], &nonce_bytes);
    const nonce_felts = bytesToU256(nonce_padded);

    return .{
        ct_felts[0],
        ct_felts[1],
        ct_felts[2],
        ct_felts[3],
        nonce_felts[0],
        nonce_felts[1],
    };
}

pub fn serializeAuditProof(proof: she.types.AuditProof) Error![11]core.Felt {
    const ax = try serializeProjectivePoint(proof.ax);
    const al0 = try serializeProjectivePoint(proof.al0);
    const al1 = try serializeProjectivePoint(proof.al1);
    const ar1 = try serializeProjectivePoint(proof.ar1);
    return .{
        ax[0],  ax[1],
        al0[0], al0[1],
        al1[0], al1[1],
        ar1[0], ar1[1],
        proof.sx, proof.sb, proof.sr,
    };
}

pub fn serializeCipherBalance(cipher: she.types.ElGamalCiphertext) Error![4]core.Felt {
    const l = try serializeProjectivePoint(cipher.l);
    const r = try serializeProjectivePoint(cipher.r);
    return .{ l[0], l[1], r[0], r[1] };
}

pub fn serializeBitProof(proof: she.types.ProofOfBit) Error![7]core.Felt {
    const a0 = try serializeProjectivePoint(proof.a0);
    const a1 = try serializeProjectivePoint(proof.a1);
    return .{ a0[0], a0[1], a1[0], a1[1], proof.c0, proof.s0, proof.s1 };
}

pub fn serializeRange(
    allocator: std.mem.Allocator,
    range: she.types.RangeProof,
) Error![]core.Felt {
    var out: std.ArrayList(core.Felt) = .empty;
    defer out.deinit(allocator);

    out.append(allocator, core.Felt.fromU64(@intCast(range.commitments.len))) catch return Error.OutOfMemory;
    for (range.commitments) |commitment| {
        const c = serializeProjectivePoint(commitment) catch return Error.InvalidPoint;
        out.append(allocator, c[0]) catch return Error.OutOfMemory;
        out.append(allocator, c[1]) catch return Error.OutOfMemory;
    }

    out.append(allocator, core.Felt.fromU64(@intCast(range.proofs.len))) catch return Error.OutOfMemory;
    for (range.proofs) |proof| {
        const p = serializeBitProof(proof) catch return Error.InvalidPoint;
        for (p) |f| {
            out.append(allocator, f) catch return Error.OutOfMemory;
        }
    }

    return out.toOwnedSlice(allocator) catch return Error.OutOfMemory;
}

pub fn serializeProofOfTransfer(
    allocator: std.mem.Allocator,
    proof: tongo_ops.ProofOfTransfer,
) Error![]core.Felt {
    var out: std.ArrayList(core.Felt) = .empty;
    defer out.deinit(allocator);

    const points = [_]core.ProjectivePoint{
        proof.a_x, proof.a_r, proof.a_r2, proof.a_b,
        proof.a_b2, proof.a_v, proof.a_v2, proof.a_bar,
    };
    for (points) |point| {
        const serialized = serializeProjectivePoint(point) catch return Error.InvalidPoint;
        out.append(allocator, serialized[0]) catch return Error.OutOfMemory;
        out.append(allocator, serialized[1]) catch return Error.OutOfMemory;
    }

    const scalars = [_]core.Felt{ proof.s_x, proof.s_r, proof.s_b, proof.s_b2, proof.s_r2 };
    for (scalars) |scalar| {
        out.append(allocator, scalar) catch return Error.OutOfMemory;
    }

    const r_aux = serializeProjectivePoint(proof.r_aux) catch return Error.InvalidPoint;
    out.append(allocator, r_aux[0]) catch return Error.OutOfMemory;
    out.append(allocator, r_aux[1]) catch return Error.OutOfMemory;

    const range = serializeRange(allocator, proof.range) catch |err| switch (err) {
        error.OutOfMemory => return Error.OutOfMemory,
        else => return Error.InvalidInput,
    };
    defer allocator.free(range);
    for (range) |f| {
        out.append(allocator, f) catch return Error.OutOfMemory;
    }

    const r_aux2 = serializeProjectivePoint(proof.r_aux2) catch return Error.InvalidPoint;
    out.append(allocator, r_aux2[0]) catch return Error.OutOfMemory;
    out.append(allocator, r_aux2[1]) catch return Error.OutOfMemory;

    const range2 = serializeRange(allocator, proof.range2) catch |err| switch (err) {
        error.OutOfMemory => return Error.OutOfMemory,
        else => return Error.InvalidInput,
    };
    defer allocator.free(range2);
    for (range2) |f| {
        out.append(allocator, f) catch return Error.OutOfMemory;
    }

    return out.toOwnedSlice(allocator) catch return Error.OutOfMemory;
}

fn bytesToU512(bytes: [64]u8) [4]core.Felt {
    var limb0: [16]u8 = undefined;
    var limb1: [16]u8 = undefined;
    var limb2: [16]u8 = undefined;
    var limb3: [16]u8 = undefined;
    @memcpy(&limb0, bytes[0..16]);
    @memcpy(&limb1, bytes[16..32]);
    @memcpy(&limb2, bytes[32..48]);
    @memcpy(&limb3, bytes[48..64]);

    return .{
        feltFromU128(std.mem.readInt(u128, &limb0, .big)),
        feltFromU128(std.mem.readInt(u128, &limb1, .big)),
        feltFromU128(std.mem.readInt(u128, &limb2, .big)),
        feltFromU128(std.mem.readInt(u128, &limb3, .big)),
    };
}

fn bytesToU256(bytes: [32]u8) [2]core.Felt {
    var high_bytes: [16]u8 = undefined;
    var low_bytes: [16]u8 = undefined;
    @memcpy(&high_bytes, bytes[0..16]);
    @memcpy(&low_bytes, bytes[16..32]);
    return .{
        feltFromU128(std.mem.readInt(u128, &low_bytes, .big)),
        feltFromU128(std.mem.readInt(u128, &high_bytes, .big)),
    };
}

fn feltFromU128(value: u128) core.Felt {
    return core.Felt.fromInt(@as(u256, value)) catch unreachable;
}

File: crates/common/src/serialization.rs
//! Cairo type serialization for TONGO contract interactions.
//!
//! This module handles conversion between Rust types and Cairo felt arrays
//! for contract calldata and response parsing.

use crate::{
    AuditProof, ElGamalCiphertext, ElGamalProof, KmsError, Poe2Proof, PoeProof, ProofOfBit,
    ProofOfTransfer, Range, Result,
};
use starknet_types_core::curve::ProjectivePoint;
use starknet_types_core::felt::Felt;

/// Serialize a ProjectivePoint to Cairo StarkPoint format `(x, y)`.
///
/// # Errors
/// Returns `KmsError::CryptoError` when the point is at infinity.
pub fn serialize_projective_point(point: &ProjectivePoint) -> Result<(Felt, Felt)> {
    let affine = point
        .to_affine()
        .map_err(|_| KmsError::CryptoError("Invalid point".to_string()))?;

    Ok((affine.x(), affine.y()))
}

/// Deserialize Cairo StarkPoint `(x, y)` to ProjectivePoint.
///
/// # Errors
/// Returns `KmsError::CryptoError` if the coordinates are not on the curve.
pub fn deserialize_projective_point(x: Felt, y: Felt) -> Result<ProjectivePoint> {
    ProjectivePoint::from_affine(x, y)
        .map_err(|_| KmsError::CryptoError("Invalid point coordinates".to_string()))
}

/// Serialize Proof of Exponentiation (PoE) proof for Cairo.
///
/// Cairo serialization: `[Ax, Ay, s]`.
pub fn serialize_poe_proof(proof: &PoeProof) -> Result<Vec<Felt>> {
    Ok(vec![proof.a.x, proof.a.y, proof.s])
}

/// Serialize Proof of Exponentiation 2 (PoE2) proof for Cairo.
///
/// Cairo serialization: `[Ax, Ay, s1, s2]`.
pub fn serialize_poe2_proof(proof: &Poe2Proof) -> Result<Vec<Felt>> {
    Ok(vec![proof.a.x, proof.a.y, proof.s1, proof.s2])
}

/// Serialize ElGamal proof for Cairo.
///
/// Cairo serialization: `[ALx, ALy, ARx, ARy, sb, sr]`.
pub fn serialize_elgamal_proof(proof: &ElGamalProof) -> Result<Vec<Felt>> {
    Ok(vec![
        proof.al.x, proof.al.y, proof.ar.x, proof.ar.y, proof.sb, proof.sr,
    ])
}

/// Convert u128 to Cairo u256 `(low, high)` representation.
#[must_use]
pub fn u128_to_u256(value: u128) -> (Felt, Felt) {
    (Felt::from(value), Felt::ZERO)
}

/// Convert u256 `(low, high)` to u128 if it fits.
///
/// # Errors
/// Returns `KmsError::CryptoError` if the value does not fit into `u128`.
pub fn u256_to_u128(low: Felt, high: Felt) -> Result<u128> {
    if high != Felt::ZERO {
        return Err(KmsError::CryptoError(
            "Value too large for u128".to_string(),
        ));
    }

    let bytes = low.to_bytes_be();
    let mut u128_bytes = [0u8; 16];
    u128_bytes.copy_from_slice(&bytes[16..32]);

    Ok(u128::from_be_bytes(u128_bytes))
}

/// Serialize AEBalance (Authenticated Encryption balance hint).
///
/// # Errors
/// Returns `KmsError::CryptoError` if the ciphertext or nonce lengths are invalid.
pub fn serialize_ae_balance(ciphertext_bytes: &[u8], nonce_bytes: &[u8]) -> Result<Vec<Felt>> {
    if ciphertext_bytes.len() != 64 {
        return Err(KmsError::CryptoError(format!(
            "Ciphertext must be 64 bytes, got {}",
            ciphertext_bytes.len()
        )));
    }
    if nonce_bytes.len() != 24 {
        return Err(KmsError::CryptoError(format!(
            "Nonce must be 24 bytes, got {}",
            nonce_bytes.len()
        )));
    }

    let ct_felts = bytes_to_u512(ciphertext_bytes);

    let mut nonce_padded = [0u8; 32];
    nonce_padded[..24].copy_from_slice(nonce_bytes);
    let nonce_felts = bytes_to_u256(&nonce_padded);

    Ok(vec![
        ct_felts.0,
        ct_felts.1,
        ct_felts.2,
        ct_felts.3,
        nonce_felts.0,
        nonce_felts.1,
    ])
}

fn bytes_to_u512(bytes: &[u8]) -> (Felt, Felt, Felt, Felt) {
    let mut limb0 = [0u8; 16];
    let mut limb1 = [0u8; 16];
    let mut limb2 = [0u8; 16];
    let mut limb3 = [0u8; 16];

    limb0.copy_from_slice(&bytes[0..16]);
    limb1.copy_from_slice(&bytes[16..32]);
    limb2.copy_from_slice(&bytes[32..48]);
    limb3.copy_from_slice(&bytes[48..64]);

    (
        Felt::from(u128::from_be_bytes(limb0)),
        Felt::from(u128::from_be_bytes(limb1)),
        Felt::from(u128::from_be_bytes(limb2)),
        Felt::from(u128::from_be_bytes(limb3)),
    )
}

fn bytes_to_u256(bytes: &[u8]) -> (Felt, Felt) {
    let mut low_bytes = [0u8; 16];
    let mut high_bytes = [0u8; 16];

    low_bytes.copy_from_slice(&bytes[16..32]);
    high_bytes.copy_from_slice(&bytes[0..16]);

    (
        Felt::from(u128::from_be_bytes(low_bytes)),
        Felt::from(u128::from_be_bytes(high_bytes)),
    )
}

/// Serialize CairoOption::Some variant.
#[must_use]
pub fn serialize_cairo_some<F>(data: F) -> Vec<Felt>
where
    F: FnOnce() -> Vec<Felt>,
{
    let mut result = vec![Felt::ZERO];
    result.extend(data());
    result
}

/// Serialize CairoOption::None variant.
#[must_use]
pub fn serialize_cairo_none() -> Vec<Felt> {
    vec![Felt::ONE]
}

/// Serialize Audit proof for Cairo.
pub fn serialize_audit_proof(proof: &AuditProof) -> Result<Vec<Felt>> {
    Ok(vec![
        proof.ax.x,
        proof.ax.y,
        proof.al0.x,
        proof.al0.y,
        proof.al1.x,
        proof.al1.y,
        proof.ar1.x,
        proof.ar1.y,
        proof.sx,
        proof.sb,
        proof.sr,
    ])
}

/// Serialize CipherBalance (ElGamal ciphertext) for Cairo.
pub fn serialize_cipher_balance(cipher: &ElGamalCiphertext) -> Result<Vec<Felt>> {
    let (l_x, l_y) = serialize_projective_point(&cipher.l)?;
    let (r_x, r_y) = serialize_projective_point(&cipher.r)?;

    Ok(vec![l_x, l_y, r_x, r_y])
}

/// Serialize ProofOfBit for Cairo.
pub fn serialize_bit_proof(proof: &ProofOfBit) -> Result<Vec<Felt>> {
    Ok(vec![
        proof.a0.x, proof.a0.y, proof.a1.x, proof.a1.y, proof.c0, proof.s0, proof.s1,
    ])
}

/// Serialize Range proof for Cairo.
pub fn serialize_range(range: &Range) -> Result<Vec<Felt>> {
    let mut felts = Vec::new();

    felts.push(Felt::from(range.commitments.len()));
    for commitment in &range.commitments {
        felts.push(commitment.x);
        felts.push(commitment.y);
    }

    felts.push(Felt::from(range.proofs.len()));
    for proof in &range.proofs {
        felts.extend(serialize_bit_proof(proof)?);
    }

    Ok(felts)
}

/// Serialize ProofOfTransfer for Cairo.
pub fn serialize_proof_of_transfer(proof: &ProofOfTransfer) -> Result<Vec<Felt>> {
    let mut felts = Vec::new();

    for point in [
        &proof.a_x,
        &proof.a_r,
        &proof.a_r2,
        &proof.a_b,
        &proof.a_b2,
        &proof.a_v,
        &proof.a_v2,
        &proof.a_bar,
    ] {
        felts.push(point.x);
        felts.push(point.y);
    }

    felts.extend([proof.s_x, proof.s_r, proof.s_b, proof.s_b2, proof.s_r2]);
    felts.extend(serialize_range(&proof.range)?);
    felts.extend(serialize_range(&proof.range2)?);

    Ok(felts)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::SerializablePoint;

    #[test]
    fn serialize_poe_proof_keeps_typed_felts() {
        let proof = PoeProof {
            a: SerializablePoint {
                x: Felt::from(1u64),
                y: Felt::from(2u64),
            },
            s: Felt::from(3u64),
            c: Felt::from(4u64),
        };

        assert_eq!(
            serialize_poe_proof(&proof).unwrap(),
            vec![Felt::from(1u64), Felt::from(2u64), Felt::from(3u64)]
        );
    }

    #[test]
    fn serialize_ae_balance_validates_sizes() {
        assert!(serialize_ae_balance(&[0u8; 63], &[0u8; 24]).is_err());
        assert!(serialize_ae_balance(&[0u8; 64], &[0u8; 23]).is_err());
    }

    #[test]
    fn serialize_roundtrip_point() {
        let g_x =
            Felt::from_hex("0x1ef15c18599971b7beced415a40f0c7deacfd9b0d1819e03d723d8bc943cfca")
                .unwrap();
        let g_y =
            Felt::from_hex("0x5668060aa49730b7be4801df46ec62de53ecd11abe43a32873000c36e8dc1f")
                .unwrap();
        let point = ProjectivePoint::from_affine(g_x, g_y).unwrap();

        let (x, y) = serialize_projective_point(&point).unwrap();
        let recovered = deserialize_projective_point(x, y).unwrap();

        assert_eq!(serialize_projective_point(&recovered).unwrap(), (g_x, g_y));
    }

    #[test]
    fn serialize_range_emits_lengths() {
        let range = Range {
            commitments: vec![SerializablePoint {
                x: Felt::from(1u64),
                y: Felt::from(2u64),
            }],
            proofs: vec![ProofOfBit {
                a0: SerializablePoint {
                    x: Felt::from(3u64),
                    y: Felt::from(4u64),
                },
                a1: SerializablePoint {
                    x: Felt::from(5u64),
                    y: Felt::from(6u64),
                },
                c0: Felt::from(7u64),
                s0: Felt::from(8u64),
                s1: Felt::from(9u64),
            }],
        };

        let felts = serialize_range(&range).unwrap();
        assert_eq!(felts[0], Felt::ONE);
        assert_eq!(felts[3], Felt::ONE);
    }
}
